File: src/main/java/mail/Mail.java
package mail;

import javax.mail.Message;
import javax.mail.MessagingException;
import javax.mail.Session;
import javax.mail.Transport;
import javax.mail.internet.InternetAddress;
import javax.mail.internet.MimeMessage;
import java.io.UnsupportedEncodingException;
import java.util.Properties;

/**
 * A utility class for sending emails.
 * @author Max Meinhold <mxmeinhold@gmail.com>
 */
public class Mail {

    /** The DEaDASS-UI host to route approval actions to. */
    private static final String HOST = null; //TODO

    private static final String REQUEST_STRING = "%s is requesting a database named %s for <code>%s</code>.<br/><br/>"
            + "<a href=%s/approvals>Visit the approvals page.</a><br/><br/>- DEaDASS";

    private static final String APPROVAL_STRING = "Your database %s has been created.<br/>"
            + " Please <a href=%s>login to DEaDASS</a> to reset the password and obtain your access credentials.<br/><br/>- DEaDASS";
    private static final String DENIAL_STRING = "Your database %s has been denied.<br/><br/>- DEaDASS";

    private static MimeMessage msgTemplate;
    private static Transport transport;


    /**
     * Initialises mail handling
     */
    public static void init() {
        Properties props;
        Session session;

        props = System.getProperties();
        props.put("mail.transport.protocol", "smtp");
        props.put("mail.smtp.port", 25);
        props.put("mail.smtp.starttls.enable", "true");
        props.put("mail.smtp.auth", "true");

        session = Session.getDefaultInstance(props);

        try {
            transport = session.getTransport();
            msgTemplate = new MimeMessage(session);
            msgTemplate.setFrom(new InternetAddress("deadass@csh.rit.edu", "DEaDASS"));
        } catch (MessagingException e) {
            System.err.println("Mail init failed.");
            e.printStackTrace();
        } catch (UnsupportedEncodingException e) {
            System.err.println("Mail return address bad.");
            e.printStackTrace();
        }
    }


    /**
     * Sends an email
     * @param subject the subject line
     * @param body the message body
     * @param uid the user to send to
     */
    private static void sendMail(String subject, String body, String uid) {
        try {
            msgTemplate.setRecipient(Message.RecipientType.TO, new InternetAddress(uid + "@csh.rit.edu"));
            msgTemplate.setSubject(subject);
            msgTemplate.setContent(body, "text/html");
            transport.connect(defaults.Secrets.MAIL_HOST, defaults.Secrets.MAIL_USERNAME, defaults.Secrets.MAIL_PASSWORD);

            transport.sendMessage(msgTemplate, msgTemplate.getAllRecipients());

            transport.close();
        } catch (MessagingException e) {
            System.err.println("Mail sendMail failed.");
            e.printStackTrace();
        }
    }


    /**
     * Sends a database request to an administrator
     * @param uid the user requesting a database
     * @param purpose the purpose of this database
     * @param dbName the name of the database
     */
    public static void request(String uid, String purpose, String dbName) {
        String subject = String.format("Database Request - %s:%s", uid, dbName);
        String body = String.format(REQUEST_STRING, uid, dbName, purpose, HOST);
        sendMail(subject, body, defaults.Secrets.MAIL_ADMIN_UID);
    }


    /**
     * Sends a notification of database approval
     * @param uid the user to notify
     * @param dbName the name of the db that was approved
     */
    public static void approve(String uid, String dbName) {
        String body = String.format(APPROVAL_STRING, dbName, HOST);
        sendMail("Database Creation", body, uid);
    }


    /**
     * Sends a notification of database denial.
     * @param uid the user to notify
     * @param dbName the name of the db that was denied.
     */
    public static void deny(String uid, String dbName) {
        String body = String.format(DENIAL_STRING, dbName);
        sendMail("Database Denial", body, uid);
    }
}

File: backend/src/mail.ts
import nodemailer = require("nodemailer");
import Mail = require("nodemailer/lib/mailer");

export type MailConfigStanza = {
  host: string;
  auth: {
    user: string;
    pass: string;
  };
  // Defaults to 25. Use 465 for secure
  port: number | undefined;
  // administrator mail address for approving requests
  admin_address: string;
  // url of the frontend, for links in mail
  ui_host: string;
};

class Mailer {
  private readonly transport: Mail;
  // Frontend url
  private readonly ui_host: string;
  // Address of people who can approve requests
  private readonly admin_address: string;

  /**
   * @param config the mail config stanza
   */
  public constructor(config: MailConfigStanza) {
    this.admin_address = config.admin_address;
    this.ui_host = config.ui_host;
    this.transport = nodemailer.createTransport({
      host: config.host,
      port: config.port || 25,
      secure: config.port == 465, // False if not port 465
      auth: config.auth,
    });
  }

  /**
   * Send an email
   * @param recipient  The address to send mail to
   * @param subject  The message subject
   * @param body  The message body
   */
  private async send(
    recipient: string,
    subject: string,
    body: string
  ): Promise<void> {
    await this.transport.sendMail({
      from: '"DEaDASS" <noreply@csh.rit.edu>',
      to: recipient,
      subject: subject,
      html: body,
    });
  }

  /**
   * Notify the administrator of a new pending request
   * @param uid  the user who is requesting a db
   * @param purpose  the stated reason for the new db
   * @param db_name  the name of the new db
   * @returns  a promise that will resolve when the email sending succeeds
   */
  public request(uid: string, purpose: string, db_name: string): Promise<void> {
    const body: string =
      `${uid} is requesting a database named ${db_name} because ` +
      `<code>${purpose}</code>,<br/><br/><a href=${this.ui_host}/approvals>` +
      `Visit the approvals page.</a><br/><br/>- DEaDASS`;
    return this.send(
      this.admin_address,
      `Database Request - ${uid}:${db_name}`,
      body
    );
  }

  /**
   * Send an approval email
   * @param uid  the user who's request has been approved
   * @param db_name  the name of the db that has been approved
   * @returns  a promise that will resolve when the email sending succeeds
   */
  public approve(uid: string, db_name: string): Promise<void> {
    const body: string =
      `Your database ${db_name} has been created.<br/>` +
      `Please <a href=${this.ui_host}>login to DEaDASS</a> to reset the ` +
      `password and obtain your access credentials.<br/><br/>- DEaDASS`;
    return this.send(
      `${uid}@csh.rit.edu`,
      `Database Creation - ${db_name}`,
      body
    );
  }

  /**
   * Send a denial email
   * @param uid  the user who's request has been denied
   * @param db_name  the name of the db that has been denied
   * @returns  a promise that will resolve when the email sending succeeds
   */
  public deny(uid: string, db_name: string): Promise<void> {
    const body = `Your database ${db_name} has been denied.<br/><br/>- DEaDASS`;
    return this.send(
      `${uid}@csh.rit.edu`,
      `Database Denial - ${db_name}`,
      body
    );
  }
}

export default Mailer;
